### PR TITLE
Zoarif/nav and details list accessibility

### PR DESCRIPTION
**Pull request by zoarif**
The pull request includes:
1) Adding an aria label prop on Nav link.
2) Adding a new callback in Selection Zone to indicate that the list item was selected via space or click.
3) Fixing the aria role on grid in details list

### CODE REVIEW

File: src/components/DetailsList/DetailsList.Props.ts [hunk 1]
```diff
@@ -134,6 +134,9 @@ export interface IDetailsListProps extends React.Props<DetailsList> {
 
   /** Boolean value to indicate if the role application should be applied on details list. Set to false by default */
   shouldApplyApplicationRole?: boolean;
+
+  /** Callback for selection zone to indicate whether the item was selected */
+  onItemSelected?: (isSelected: boolean) => void;
```
Comment (dzearing) [inline src/components/DetailsList/DetailsList.Props.ts]: this doesn't make sense to me. What is the use case here? Why would a user want this?

Comment (zoarif) [inline src/components/DetailsList/DetailsList.Props.ts]: The main idea behind that was there was no indication is selection zone whether the item was selected using spacebar or click. I wanted to know that because we want to add an aria alert when the list item gets selected.

Comment (antonlabunets) [inline src/components/DetailsList/DetailsList.Props.ts]: Can you use aria-selected instead of having to provide this?

File: src/components/Nav/Nav.tsx [hunk 2]
```diff
@@ -44,13 +44,15 @@ export class Nav extends React.Component<INavProps, INavState> {
...
+    let { onLinkClick, shouldAssignTitle = true } = this.props;
     return (
       <li key={ linkIndex }>
         <a
           className={'ms-Nav-link' + (_isLinkSelected(link) ? ' is-selected' : '')}
           href={ link.url || 'javascript:' }
-          onClick={ this.props.onLinkClick }
-          title={ link.name }
+          onClick={ onLinkClick }
+          title={ shouldAssignTitle ? link.name : '' }
```
Comment (antonlabunets) [inline src/components/Nav/Nav.tsx]: Did you confirm this skips adding title altogether or makes for an empty tooltip? If the latter, consider using undefined in the else.

File: src/demo/pages/DetailsListPage/examples/DetailsList.Advanced.Example.tsx [hunk 3]
```diff
@@ -393,6 +395,10 @@ export class DetailsListAdvancedExample extends React.Component<any, IDetailsLis
     console.log('Item invoked', item, index);
   }
 
+  private _onItemSelected(isSelected: boolean) {
+    console.log('item', isSelected);
```
Comment (antonlabunets) [inline src/demo/pages/DetailsListPage/examples/DetailsList.Advanced.Example.tsx]: Would be best to output the index to confirm the corresponding item is selected.

Comment (zoarif) [inline src/demo/pages/DetailsListPage/examples/DetailsList.Advanced.Example.tsx]: ok

File: src/components/Nav/Nav.tsx (hunk 2 above)
Comment (zoarif) [inline src/components/Nav/Nav.tsx]: ok

File: src/components/DetailsList/DetailsList.Props.ts (hunk 1 above)
Comment (zoarif) [inline src/components/DetailsList/DetailsList.Props.ts]: We wanted to say something like this read “File Name, File Type, selected” , aria-selected doesn't seem to serve the whole purpose.

File: src/components/DetailsList/DetailsList.Props.ts [hunk 4]
```diff
@@ -134,6 +134,9 @@ export interface IDetailsListProps extends React.Props<DetailsList> {
 
   /** Boolean value to indicate if the role application should be applied on details list. Set to false by default */
   shouldApplyApplicationRole?: boolean;
+
+  /** Callback for selection zone to indicate whether the item was selected */
+  onItemSelected?: (item: any, index: any, isSelected: boolean) => void;
```
Comment (dzearing) [inline src/components/DetailsList/DetailsList.Props.ts]: Is this only called on single item selects? What about multiple?

File: src/components/DetailsList/DetailsList.tsx [hunk 5]
```diff
@@ -206,53 +207,56 @@ export class DetailsList extends React.Component<IDetailsListProps, IDetailsList
...
             isInnerZoneKeystroke={ (ev) => (ev.which === getRTLSafeKeyCode(KeyCodes.right)) }
             onActiveElementChanged={ this._onActiveRowChanged }
-          >
+            >
             <SelectionZone
               selection={ selection }
               selectionMode={ selectionMode }
-              onItemInvoked={ onItemInvoked }>
+              onItemInvoked={ onItemInvoked }
+              onItemSelected= { onItemSelected }>
```
Comment (dzearing) [inline src/components/DetailsList/DetailsList.tsx]: nit space

File: src/components/Nav/Nav.Props.ts [hunk 6]
```diff
@@ -21,6 +21,11 @@ export interface INavProps {
    * Indicates whether the navigation component renders on top of other content in the UI
    */
   isOnTop?: boolean;
+
+  /**
+   * Indicates whether the title should be assigned to the nav links
+   */
+  shouldAssignTitle?: boolean;
```
Comment (dzearing) [inline src/components/Nav/Nav.Props.ts]: This isn't very clear; what does "assign" mean?

File: src/utilities/selection/SelectionZone.tsx [hunk 7]
```diff
@@ -196,6 +198,7 @@ export class SelectionZone extends React.Component<ISelectionZoneProps, any> {
             selection.setAllSelected(false);
           }
           selection.setIndexSelected(index, !isSelected, true);
+          onItemSelected(selection.getItems()[index], index, !isSelected);
```
Comment (dzearing) [inline src/utilities/selection/SelectionZone.tsx]: i don't think this is right.

selection object itself is something that can be externallymanipulated outside of selection zone. The caller should use the selection object to get selection changes. 

A better solution here would be: expose the selection object publicly via getter, make it part of the IDetailsList contract. Then make sure the selection object has the hooks you need to observe changes.

Consider this; i have a MarqueeSelection component coming. It takes in a selection object, and enables you to drag a box around items that are selected. When it wants to set something selected, it interacts with Selection, not SelectionZone. It will say selection.setSelected(index, true) on something. Then a change event should fire and the consumer should deal with that.

File: src/components/Nav/Nav.Props.ts (hunk 6 above)
Comment (zoarif) [inline src/components/Nav/Nav.Props.ts]: indicates whether the nav link should be given a title

**Comment by dzearing** [comment]
Approved

File: src/components/DetailsList/DetailsList.Props.ts (hunk 1 above)
Comment (dzearing) [inline src/components/DetailsList/DetailsList.Props.ts]: I don't want to add this callback. Instead this should be a selection thing.

File: src/utilities/selection/Selection.ts [hunk 8]
```diff
@@ -13,9 +13,11 @@ export class Selection implements ISelection {
   private _keyToIndexMap: { [key: string]: number };
   private _anchoredIndex: number;
   private _onSelectionChanged: () => void;
+  private _onItemSelected: (item?: any, index?: number, isSelected?: boolean) => void;
   private _hasChanged: boolean;
 
-  constructor(onSelectionChanged?: () => void, getKey?: (item: IObjectWithKey, index?: number) => string) {
+  constructor(onSelectionChanged?: () => void, getKey?: (item: IObjectWithKey, index?: number) => string,
+  onItemSelected?: (item?: any, index?: number, isSelected?: boolean) => void) {
```
Comment (dzearing) [inline src/utilities/selection/Selection.ts]: i'm a little confused; why is this needed when the first param is onSelectionChanged? isn't the onSelectionChanged callback enough?
